DataDog/dd-trace-dotnet
# Automatically instrument methods decorated with `[Trace]` (`Datadog.Trace.Annotations.TraceAttribute`)

## Summary of changes
Adds a new public type named `Datadog.Trace.Annotations.TraceAttribute`. Application code can add the attribute to any method with `[Trace]` and automatic instrumentation will automatically instrument it at runtime, with the following span metadata.

To avoid having applications depend on `Datadog.Trace.dll` for this attribute, define it in a new assembly `Datadog.Trace.Annotations.dll`.

### Span metadata
Operation Name: `TraceAttribute.OperationName ?? "trace.annotation"`
Resource Name: `TraceAttribute.ResourceName ?? "<METHOD_NAME>"`
Tags:
- component: trace

### Configuration
This feature is enabled by default. If the overhead is too much and users are not using the feature, it can be disabled by setting `DD_TRACE_ANNOTATIONS_ENABLED=false`.

## Implementation details
- Add a new PInvoke called `AddTraceAttributeInstrumentation` which is invoked early (like the ByRef instrumentation) before the static `Tracer.Instance` is initialized. This initializes the `TraceAnnotationsIntegration` used by both `[Trace]` instrumentation and `DD_TRACE_METHODS` instrumentation.
- Modify the ModuleLoadFinished event in two major ways
  - After checking that a module should not be skipped, find any methods within the module decorated with the CustomAttribute named `Datadog.Trace.Annotations.TraceAttribute`. If the `TraceAnnotationsIntegration` has been initialized, immediately request a ReJIT for each method with integration_type=`TraceAnnotationsIntegration`. If the integration is not yet initialized, put all the found methods on a queue to be processed in a later ModuleLoadFinished callback.
  - Regardless if a module requested ReJIT, at the end of each ModuleLoadFinished callback check whether the `TraceAnnotationsIntegration` has been initialized, and if it is then empty the queue and request a ReJIT for each method.

## Test coverage
Update Samples.TraceAnnotations test application and its integration test in the following ways:
- Remove `Samples.TraceAnnotations.Program[RunTestsAsync]` from the `DD_TRACE_METHODS` configuration and place the `[Trace]` attribute on the method and assert that the span is still generated
- Add `[Trace]` attributes onto existing methods specified by `DD_TRACE_METHODS` and assert that the spans have the same OperationName/ResourceName as the attribute
- Add a new methods that only have the `[Trace]` attribute and are not specified in `DD_TRACE_METHODS`. Assert that new spans are created for the new methods
- Use both the official `Datadog.Trace.Annotatations.TraceAttribute` type and a user-defined `Datadog.Trace.Annotations.TraceAttribute` to ensure that both result in instrumentation
